chia/rpc/wallet_rpc_api.py

```diff
@@ -4496,3 +4504,83 @@ class CRCATApprovePending(Streamable):
+    @tx_endpoint(push=True)
+    async def vault_create(
+        self,
+        request: Dict[str, Any],
+        tx_config: TXConfig = DEFAULT_TX_CONFIG,
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> EndpointResult:
+        """
+        Create a new vault
+        """
+        assert self.service.wallet_state_manager
+        wallet = self.service.wallet_state_manager.main_wallet
+
+        secp_pk = bytes.fromhex(str(request.get("secp_pk")))
+        entropy = bytes.fromhex(str(request.get("entropy")))
+        bls_pk = G1Element.from_bytes(bytes.fromhex(str(request.get("bls_pk"))))
+        timelock = uint64(request["timelock"])
+        fee = request.get("fee", 0)
+        vault_puzzlehash = get_vault_puzzle_hash(
+            secp_pk, DEFAULT_CONSTANTS.GENESIS_CHALLENGE, entropy, bls_pk, timelock
+        )
+
+        # Get xch coin
+        amount = uint64(1)
+        coins = await wallet.select_coins(uint64(amount + fee), tx_config.coin_selection_config)
+
+        # Create singleton launcher
+        origin = coins.copy().pop()
+        launcher_coin = Coin(origin.name(), SINGLETON_LAUNCHER_PUZZLE_HASH, amount)
+
+        vault_full_puzzlehash = create_singleton_puzzle_hash(vault_puzzlehash, launcher_coin.name())
+        announcement_message = Program.to([vault_puzzlehash, amount, bytes(0x80)]).get_tree_hash()
+
+        [tx_record] = await wallet.generate_signed_transaction(
+            amount,
+            SINGLETON_LAUNCHER_PUZZLE_HASH,
+            tx_config,
+            fee,
+            coins,
+            None,
+            origin_id=origin.name(),
+            extra_conditions=(
+                AssertCoinAnnouncement(asserted_id=launcher_coin.name(), asserted_msg=announcement_message),
+            ),
+        )
+
+        genesis_launcher_solution = Program.to([vault_puzzlehash, amount, bytes(0x80)])
+
+        launcher_cs = CoinSpend(launcher_coin, SINGLETON_LAUNCHER_PUZZLE, genesis_launcher_solution)
+        launcher_sb = SpendBundle([launcher_cs], AugSchemeMPL.aggregate([]))
+        assert tx_record.spend_bundle is not None
+        full_spend = SpendBundle.aggregate([tx_record.spend_bundle, launcher_sb])
+
+        vault_record = TransactionRecord(
+            confirmed_at_height=uint32(0),
+            created_at_time=uint64(int(time.time())),
+            amount=uint64(amount),
+            to_puzzle_hash=vault_full_puzzlehash,
+            fee_amount=fee,
+            confirmed=False,
+            sent=uint32(0),
+            spend_bundle=full_spend,
+            additions=full_spend.additions(),
+            removals=full_spend.removals(),
+            wallet_id=wallet.id(),
+            sent_to=[],
+            trade_id=None,
+            type=uint32(TransactionType.INCOMING_TX.value),
+            name=full_spend.name(),
+            memos=[],
+            valid_times=ConditionValidTimes(),
+        )
```

Comment:
I'd like to keep the RPCs as pass-thru as possible so maybe this should all go into WSM